File: simple-stack-example-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/util/BasePresenter.java
package com.zhuinden.simplestackdemoexamplefragments.util;

import com.zhuinden.simplestack.Bundleable;

/**
 * Created by Owner on 2017. 01. 27..
 */

public abstract class BasePresenter<F extends BaseFragment<F, P>, P extends BasePresenter<F, P>>
        implements Bundleable {
    F fragment;

    public F getFragment() {
        return fragment;
    }

    public final void attachFragment(F fragment) {
        this.fragment = fragment;
        onAttach(fragment);
    }

    public final void detachFragment(F fragment) {
        onDetach(fragment);
        this.fragment = null;
    }

    protected abstract void onAttach(F fragment);

    protected abstract void onDetach(F fragment);
}

File: simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/util/BasePresenter.kt
package com.zhuinden.simplestackdemoexamplefragments.util

import com.zhuinden.simplestack.Bundleable

/**
 * Created by Zhuinden on 2018. 08. 20.
 */

abstract class BasePresenter<V : BaseViewContract> : Bundleable {
    var view: V? = null

    fun attachFragment(fragment: V) {
        this.view = fragment
        onAttach(fragment)
    }

    fun detachFragment(fragment: V) {
        onDetach(fragment)
        this.view = null
    }

    protected abstract fun onAttach(view: V)

    protected abstract fun onDetach(view: V)
}
